### PR TITLE
[MI-2075] Add test cases for oAuth

### PR DESCRIPTION
**Pull request by ayusht2810**

### CODE REVIEW

File: server/plugin/oAuth_test.go [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+			monkey.PatchInstanceMethod(reflect.TypeOf(&p), "DM", func(_ *Plugin, _, _ string, _ ...interface{}) (string, error) {
+				return "", testCase.DMErr
+			})
+
+			req := httptest.NewRequest(http.MethodGet, "/oauth/connect", bytes.NewBufferString(`{}`))
+			req.Header.Add(constants.HeaderMattermostUserID, testCase.mattermostUserID)
+
+			res := httptest.NewRecorder()
+
+			p.OAuthConnect(res, req)
```
Comment (raghavaggarwal2308) [inline server/plugin/oAuth_test.go:71]: check for something

File: server/plugin/oAuth_test.go [hunk 2]
```diff
@@ -0,0 +1,404 @@
...
+			monkey.PatchInstanceMethod(reflect.TypeOf(&p), "CloseBrowserWindowWithHTTPResponse", func(_ *Plugin, _ http.ResponseWriter) {})
+
+			req := httptest.NewRequest(http.MethodGet, "/oauth/complete", bytes.NewBufferString(`{}`))
+			q := req.URL.Query()
+			q.Add("code", testCase.code)
+			q.Add("state", testCase.state)
+			req.URL.RawQuery = q.Encode()
+			res := httptest.NewRecorder()
+
+			p.OAuthComplete(res, req)
```
Comment (raghavaggarwal2308) [inline server/plugin/oAuth_test.go:130]: same

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

**Review by raghavaggarwal2308 — APPROVED** [review]

File: server/plugin/oAuth.go [hunk 3]
```diff
@@ -108,11 +110,15 @@ func (p *Plugin) OAuthComplete(w http.ResponseWriter, r *http.Request) {
...
 
-	p.closeBrowserWindowWithHTTPResponse(w)
+	p.CloseBrowserWindowWithHTTPResponse(w)
 }
 
 // GenerateOAuthToken generates OAuth token after successful authorization
 func (p *Plugin) GenerateOAuthToken(code, state string) error {
+	if len(strings.Split(state, "_")) != 2 || strings.Split(state, "_")[1] == "" {
+		return errors.New("invalid state")
+	}
```
Comment (avas27JTG) [inline server/plugin/oAuth.go:120]: why is this added here?
we are already checking this inside `OAuthComplete` on line `103`

File: server/plugin/oAuth_test.go [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+			statusCode:  http.StatusUnauthorized,
+		},
+		{
+			description:      "test OAuthConnect with user already connected",
+			isConnected:      true,
+			mattermostUserID: "mockMattermostUserID",
+			statusCode:       http.StatusBadRequest,
+		},
+		{
+			description:      "test OAuthConnect with user already connected and failed to DM",
```
Comment (avas27JTG) [inline server/plugin/oAuth_test.go:52]: keep description consistent

File: server/plugin/oAuth_test.go [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+			state:       "mock_State",
+			statusCode:  http.StatusBadRequest,
+		},
+		{
+			description: "test OAuthComplete without state",
+			code:        "mockCode",
+			statusCode:  http.StatusBadRequest,
+		},
+		{
+			description: "test OAuthComplete with length of state not equal to 2",
```
Comment (avas27JTG) [inline server/plugin/oAuth_test.go:111]:
```suggestion
			description: "OAuthComplete: length of state not equal to 2",
```

File: server/plugin/oAuth_test.go [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+		expectedError    string
+		DMError          error
+	}{
+		{
+			description: "test GenerateOAuthToken",
+			code:        "mockCode",
+			state:       "mock_state",
+		},
+		{
+			description:   "test GenerateOAuthToken without state",
```
Comment (avas27JTG) [inline server/plugin/oAuth_test.go:171]:
```suggestion
			description:   "GenerateOAuthToken: without state",
```

File: server/plugin/oAuth_test.go [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+			expectedError: "mockError",
+		},
+		{
+			description:   "test RefreshOAuthToken when user is not loaded successfully and DM error occurs",
+			loadUserErr:   errors.New("mockError"),
+			DMErr:         errors.New("mockError"),
+			expectedError: "mockError",
+		},
+		{
+			description:   "test RefreshOAuthToken when token is not decoded successfully",
```
Comment (avas27JTG) [inline server/plugin/oAuth_test.go:236]:
```suggestion
			description:   "RefreshOAuthToken: token is not decoded successfully",
```

File: server/plugin/oAuth.go (hunk 3 above)
Comment (ayusht2810) [reply server/plugin/oAuth.go]: reverted

**Review by avas27JTG — APPROVED** [review]
